numpyro/infer/util.py

```diff
@@ -168,84 +201,84 @@ def body_fn(state):
         key, subkey = random.split(key)
 
-        # Wrap model in a `substitute` handler to initialize from `init_loc_fn`.
-        seeded_model = substitute(seed(model, subkey), substitute_fn=init_strategy)
-        model_trace = trace(seeded_model).get_trace(*model_args, **model_kwargs)
-        constrained_values, inv_transforms = {}, {}
-        for k, v in model_trace.items():
-            # TODO: filter out discrete sites here
-            if v['type'] == 'sample' and not v['is_observed']:
-                constrained_values[k] = v['value']
-                inv_transforms[k] = biject_to(v['fn'].support)
-        params = transform_fn(inv_transforms,
-                              {k: v for k, v in constrained_values.items()},
-                              invert=True)
-        potential_fn = partial(potential_energy, model, inv_transforms, model_args, model_kwargs)
-        pe, param_grads = value_and_grad(potential_fn)(params)
-        z_grad = ravel_pytree(param_grads)[0]
-        is_valid = np.isfinite(pe) & np.all(np.isfinite(z_grad))
-        return i + 1, key, params, is_valid
-
-    def _find_valid_params(rng_key_):
-        _, _, prototype_params, is_valid = init_state = body_fn((0, rng_key_, None, None))
-        # Early return if valid params found.
-        if not_jax_tracer(is_valid):
-            if device_get(is_valid):
-                return prototype_params, is_valid
-
-        # TODO: we are tracing the model 4 times here, make sure that we just do it one
-        # Historically, we don't know the initial structure of params
-        # so we need the do-while pattern (lax.while_loop requires initial values)
-        # In addition, the body_fn requires tracing through model 2 times: 1 to find the params
-        # the other one is to compute potential_energy...
-        # TODO: add tests to track down the number of time we trace the model
-        # and try to optimize it in follow-up PRs
-        _, _, init_params, is_valid = while_loop(cond_fn, body_fn, init_state)
-        return init_params, is_valid
+        if radius is None or prototype_params is None:
+            # Wrap model in a `substitute` handler to initialize from `init_loc_fn`.
+            seeded_model = substitute(seed(model, subkey), substitute_fn=init_strategy)
+            model_trace = trace(seeded_model).get_trace(*model_args, **model_kwargs)
+            constrained_values, inv_transforms = {}, {}
+            for k, v in model_trace.items():
+                # TODO: filter out discrete sites here
+                if v['type'] == 'sample' and not v['is_observed']:
+                    constrained_values[k] = v['value']
+                    inv_transforms[k] = biject_to(v['fn'].support)
+            params = transform_fn(inv_transforms,
+                                  {k: v for k, v in constrained_values.items()},
+                                  invert=True)
+        else:  # this branch doesn't require tracing the model
```

Comment:
For `init_to_uniform`, `init_to_value`, we don't need to trace the model to get initial values. We handle that logic here.